engine/utils/HeapArray.h

```diff
@@ -0,0 +1,735 @@
+         */
+        T& operator*()
+        {
+            return *pointer;
+        }
+
+        /**
+         * @brief Increments the pointer and index
+         */
+        void Increment()
+        {
+            pointer++;
+            index++;
+        }
+
+        /**
+         * @brief Checks the next index to see if it's active.
+         * @return `true` if the next index is inactive, `false` if it is.
+         */
+        bool IsNextIndexInvalid()
+        {
+            // Get the chunk of bytes we want to use to search for states.
+            uint8_t byteMask = maskPointer[(index + 1) / BYTE_SIZE_IN_BITS];
+
+            /**
+             * Get a bit representation of the position we want to search for.
+             * For example, given a bit position in index 1, we create a bit mask with a value of
+             * 1 left shifted by 1 position (0000 0010). Finally, since our bytes contain 8 bits,
+             * we want to make sure that we just check the index we get is relative to our byte.
+             * i.e: 2 % 8 = 2 or 9 % 8 = 1. No matter how large the index, we'll always get its
+             * position in 8s.
+             */
+            auto bitPosition = (1 << ((index + 1) % BYTE_SIZE_IN_BITS));
+
+            /**
+             * Run a bitwise AND operation to check if the corresponding byte exists in the original
+             * mask. i.e: 1010 0010 & 0000 0010 will return 1 (since 1 exists in the second bit).
+             *
+             * Finally, we also want to make sure the next index is in bounds.
+             */
+            return (byteMask & bitPosition) == 0 && (index + 1) < size;
+        }
+
+        /**
+         * @brief An equal operator, checks with equality.
+         * @param other the other iterator to compare to.
+         * @return `true` if the pointers stored by both iterators are the same.
+         */
+        bool operator==(const Iterator& other) const
+        {
+            return pointer == other.pointer;
+        }
+
+        /**
+         * @brief A non equality operator. Checks that twio iterators are not equal.
+         * @param other the iterator to compare to.
+         * @return `true` if the pointer addresses stored by both iterators are not the same.
+         */
+        bool operator!=(const Iterator& other) const
+        {
+            return pointer != other.pointer;
+        }
+
+    private:
+
+        T* pointer;
+        uint8_t* maskPointer;
+        size_t index;
+        size_t size;
+    };
+
+    /**
+     * @brief An empty constructor. This is invoked when no parameters are passed to the
+     * constructor. Initialises the array with a size of zero. Does not allocate any space on the
+     * heap. Constructing an array this way requires either an explicit resizing or a new creation
+     * to add a size. Any usage of the Get() or subscript operator will cause an error.
+     */
+    HeapArray() : HeapArray(0, 0, 0, nullptr, nullptr) {}
```

Comment:
Is there any real use-case for this constructor?

Reply:
If you want to create an empty HeapArray which will be used later. I use this pattern a few times in the renderer.